### DISCUSSION TITLE
[SOLVED] Full Pool event on fixed pools

### DISCUSSION THREAD

**Discussion by pioardi** [poolifier/poolifier#180] [General]
The idea here is to emit the event FullPool also on the fixed pools, the event will be emitted when all the workers are busy and a new task is submitted.

@jerome-benoit @Shinigami92

**Comment by Shinigami92** [comment]
I wouldn't call it `'FullPool'` but `'busy'`

Also I would prefer to rely on a concise naming convention only using `lowercase` because NodeJS does it this way
e.g. Worker from `worker_threads` or Worker from `cluster`

`'busy'` could also have the same meaning as "full" because it means "All workers that could be available are currently working"
Were "full" doesn't make really sense for a fixed pool, because it is always "full" but not "busy"

**Reply by pioardi** [reply]
Yes I agree

**Reply by pioardi** [reply]
Then if also @jerome-benoit  agree, we can create an enhancement with those details:

- Change FullPool event to busy ( or something similar, busy is ok for me )
- Add the busy event also the the fixed pools ( that means basically move a piece of code from dynamic to abstract )

**Reply by jerome-benoit** [reply]
You want to emit that event when all workers have at least one task running? 
How you will track the task statuses?

**Reply by pioardi** [reply]
So we already track the Number of task in progress for each worker ( the tasks Map).

So we only Need to use It even for fixed pools.

And yes when all the workers have at least One Active task and a new task comes into the pull we Will emit an event

**Reply by jerome-benoit** [reply]
How will you guarantee that the current worker task is running and not in a transient state? 
But I agree, I'm just nitpicking to see if the inherent race conditions that are in event notifications can have a real world impact for poolifier and if locking should be introduced to deal with them..

**Reply by pioardi** [reply]
So our workers Just perform One kind of tasks and this Is basically how we ensure that Is running and not transient ( yes could be some small period of time before really start to work on that).

Consider that we already emit this event, but only for Dynamic pools.

**Reply by pioardi** [reply]
I tried to draft this but since that this could introduce a ( even if small ) performance overhead I will hold on for now.

**Reply by pioardi** [reply]
I think we really need continuous benchmarking versus other pools as next steps.
I would leave any other small feature for later.

**Reply by jerome-benoit** [reply]
I think the implementation of the 'busy' event sending when all workers in a pool have at least one task running will make the worker strategies selection code even more self contained: it currently sends that event inside a the choose method and it should not. 

+ Fixed pool: the code must look for task numbers on each worker at task submission and emit the 'busy' event if number of worker with a task = pool size
+ Dynamic pool: the code must look for the number of workers at worker creation and emit the 'busy' event if number of workers = pool max size

I think that logic can be shared and factorised between pools but we have to make sure it will not duplicate, for performance reason, the logic in dynamic pool that trigger the creation of a worker. 

Since it's a breaking change and a desirable code logic modularisation, I think it should be done before the version 2 release.
